### PR TITLE
Make `GuiHandle::getNode()` return a `<T extends Node> T` #219

### LINKED ISSUE #219: Make `GuiHandle::getNode()` return a `<T extends Node> T`
It's slightly undesirable that [`GuiHandle::getNode()`](https://github.com/se-edu/addressbook-level4/blob/311ab3c0ccd4d6e49d81234e6d8fe42646e3c834/src/test/java/guitests/guihandles/GuiHandle.java#L53) returns a `Node`, since it means that we need to add down-casts in order to get the subclass of `Node` that we want.

The PR should also remove the casts at callsites that were previously due to `getNode()` returning a `Node`.

### PR DESCRIPTION
**Pull request by MightyCupcakes**
Resolves #219

### CODE REVIEW

File: src/test/java/guitests/guihandles/GuiHandle.java [hunk 1]
```diff
@@ -50,17 +50,20 @@ public void focusOnWindow(String stageTitle) {
         logger.info("Finishing focus " + stageTitle);
     }
 
-    protected Node getNode(String query) {
-        return guiRobot.lookup(query).tryQuery().get();
+    @SuppressWarnings("unchecked")
+    protected <T extends Node> T getNode(String query) {
+        return (T) guiRobot.lookup(query).tryQuery().get();
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/GuiHandle.java]: Shouldn't `guiRobot.lookup(query).query()` be sufficient?

File: src/test/java/guitests/guihandles/GuiHandle.java [hunk 2]
```diff
@@ -50,17 +50,20 @@ public void focusOnWindow(String stageTitle) {
         logger.info("Finishing focus " + stageTitle);
     }
 
-    protected Node getNode(String query) {
-        return guiRobot.lookup(query).tryQuery().get();
+    @SuppressWarnings("unchecked")
+    protected <T extends Node> T getNode(String query) {
+        return (T) guiRobot.lookup(query).query();
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/GuiHandle.java]: I think you can remove the cast as well.

File: src/test/java/guitests/guihandles/GuiHandle.java [hunk 3]
```diff
@@ -50,17 +50,20 @@ public void focusOnWindow(String stageTitle) {
         logger.info("Finishing focus " + stageTitle);
     }
 
-    protected Node getNode(String query) {
-        return guiRobot.lookup(query).tryQuery().get();
+    @SuppressWarnings("unchecked")
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/GuiHandle.java]: This can be removed.

File: src/test/java/guitests/guihandles/GuiHandle.java [hunk 4]
```diff
@@ -73,7 +76,8 @@ protected String getTextFromLabel(String fieldId, Node parentNode) {
     }
 
     protected String getTextFromLabel(String fieldId) {
-        return ((Label) guiRobot.lookup(fieldId).tryQuery().get()).getText();
+        Label label = getNode(fieldId);
+        return label.getText();
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/GuiHandle.java]: Re: code coverage. Turns out this method is not used at all. I think we can do some cleanup by removing this method.

**Comment by pyokagan** [comment]
@MightyCupcakes The name of the PR needs to match the issue title.

**Comment by damithc** [comment]
Thanks for the contribution @MightyCupcakes 👍 .
You may also create a PR to add yourself to the list of contributors if you haven't done so already.

**Pull request merged**